### PR TITLE
remove register keyword as not cpp17 complient

### PR DESCRIPTION
**Pull request by AndreyMlashkin**
Specify library name and version:  **gperf/3.1**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by intelligide** [comment]
The recipe should not change the sources in this way. If the library does not support C++17, it is better to patch the upstream than to modify the sources. For more information, see [this discussion (#3951)](https://github.com/conan-io/conan-center-index/pull/3951/files#diff-8aa07bf0dad3063ca38d5b842b934aa4f6d59bdab84855edc30d54e333c05409).

File: recipes/gperf/all/conanfile.py [hunk 1]
```diff
@@ -59,6 +59,11 @@ def _build_configure(self):
             autotools.make()
 
     def build(self):
+        if tools.valid_min_cppstd(self, "17"):
+            tools.replace_in_file(os.path.join(self._source_subfolder, "lib", "getline.cc"),
+                                  "register int c = getc (stream);",
+                                  "int c = getc (stream);")
```
Comment (madebr) [inline recipes/gperf/all/conanfile.py:65]: I think you can create a patch and apply this change unconditionally.
A compiler should be smart enough to put a hot variable in a register.

It looks like upstream already fixed it:
https://git.savannah.gnu.org/gitweb/?p=gperf.git;a=blob;f=lib/getline.cc;h=2d97644d859849c608e6653add0228183434d2fa;hb=HEAD#l58

Comment (AndreyMlashkin) [reply recipes/gperf/all/conanfile.py]: Ok, I have moved my fix to a patch

**Review by madebr — DISMISSED** [review]

**Review by ericLemanissier — CHANGES_REQUESTED** [review]
the patch needs to be added to conandata.yml, and tools.patch must be called, otherwise nothing changes here

**Comment by ericLemanissier** [comment]
AFAIU, we are exactly in the case described in https://github.com/conan-io/conan-center-index/pull/3951/files#diff-8aa07bf0dad3063ca38d5b842b934aa4f6d59bdab84855edc30d54e333c05409R21-R24 : the package does not compile std=C++17, because the register keyword was removed from the language. Nonetheless, the patch should be submitted upstream.

**Comment by madebr** [comment]
I think the patch is the less of all evil.
You could conditionally lower the `-std=c++` standard but that would require custom code in the recipe.
A patch is an ideal way to change something once, and forget about it for the following releases.

**Review by ericLemanissier — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Comment by wdobbe** [comment]
This PR doesn't work: the name of the patch file in conandata.yml is incorrect (probably copy/paste error).
This was not discovered because the source() method of the conanfile.py misses the code to apply the patches.

I will submit a new PR to correct.

**Pull request merged**